src/mlpack/methods/ann/layer/nearest_interpolation_impl.hpp

```diff
@@ -0,0 +1,161 @@
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+NearestInterpolation<InputDataType, OutputDataType>::
+  NearestInterpolation():
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  batchSize(0)
+  {
+    // Nothing to do here.
+  }
+
+template<typename InputDataType, typename OutputDataType>
+NearestInterpolation<InputDataType, OutputDataType>::
+NearestInterpolation(const size_t inRowSize,
+                     const size_t inColSize,
+                     const size_t outRowSize,
+                     const size_t outColSize,
+                     const size_t depth) :
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth),
+    batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    batchSize = input.n_cols;
+    if (output.is_empty())
+      output.set_size(outRowSize * outColSize * depth, batchSize);
+    else
+    {
+      assert(output.n_rows == outRowSize * outColSize * depth);
+      assert(output.n_cols == batchSize);
+    }
+
+    assert(inRowSize >= 2);
+    assert(inColSize >= 2);
+
+    arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+        inRowSize, inColSize, depth * batchSize, false, false);
+    arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
+        depth * batchSize, false, true);
+
+    double scaleRow = (double) inRowSize / (double) outRowSize;
+    double scaleCol = (double) inColSize / (double) outColSize;
+
+    for (size_t i = 0; i < outRowSize; ++i)
+    {
+      const size_t rOrigin = std::floor(i * scaleRow);
+
+      for (size_t j = 0; j < outColSize; ++j)
+      {
+        const size_t cOrigin = std::floor(j * scaleCol);
+
+        for (size_t k = 0; k < depth * batchSize; ++k)
+        {
+          outputAsCube(i, j, k) = inputAsCube.slice(k)(
+              rOrigin, cOrigin);
+        }
+      }
+    }
+  }
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Backward(
+  const arma::Mat<eT>& /*input*/,
+    const arma::Mat<eT>& gradient,
+    arma::Mat<eT>& output)
+  {
+    if (output.is_empty())
+    {
+      output.set_size(inRowSize * inColSize * depth, batchSize);
+    }
+    else
+    {
+      assert(output.n_rows == inRowSize * inColSize * depth);
+      assert(output.n_cols == batchSize);
+    }
+
+    assert(outRowSize >= 2);
+    assert(outColSize >= 2);
+
+    arma::cube outputAsCube(output.memptr(), inRowSize, inColSize,
+        depth * batchSize, false, true);
+    arma::cube gradientAsCube(((arma::Mat<eT>&) gradient).memptr(), outRowSize,
+        outColSize, depth * batchSize, false, false);
```

Comment:
This is undefined, when the `output` or `gradient` matrix isn't initialized.

Reply:
You could use `n_elem` or `is_empty` to check the matrix size and initialize the matrix accordingly. Also you have to use zeros for the output matrix, to make sure it's initialized correctly.

Reply:
but the gradient is const.

Reply:
Here you use the advanced constructor to create the matrix:

```c++
arma::cube gradientAsCube(((arma::Mat<eT>&) gradient).memptr(), outRowSize,
        outColSize, depth * batchSize, false, false);
```

What this does is initialize a matrix at  `gradient.memptr()` with the size, but since `gradient` is empty, you trying to initialize it using space that isn't allocated, all sort of strange things can happen here, you could actually write in space that was initialized by some other application. So you have to set size of gradient before, since `gradient` is const you have to do it outside.

Reply:
I am a bit confused.  I looked at the  bilinear_interpolation  code. There also this same thing is being used but it  doesn't give error.
https://github.com/mlpack/mlpack/blob/1fa385493a679809822f92151add8bc50a6f3260/src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp#L132

https://github.com/mlpack/mlpack/blob/1fa385493a679809822f92151add8bc50a6f3260/src/mlpack/tests/ann_layer_test.cpp#L2134

Reply:
Also this PR was passing the build and test cases. But after a couple of syntax changes of line limit, it started to fail.

Reply:
@zoq  Moreover this should not cause the build to fail. Also in the test cases the gradient is initialized before.

Reply:
For the backward pass you just pass an instantiated matrix: `layer1.Backward(output1, output1, unzoomedOutput1);` of size 0, and inside `Backward()` you just use the pointer of an empty matrix and instantiate a matrix with size  `outColSize, depth * batchSize`, that is undefined behavior, so sometimes the test can pass sometimes the test will fail. If you use the layer in combination with the `FFN` class, the gradient matrix is initialized.

Reply:
@zoq I get it. But this PR is failing in the build itself.

Reply:
Both nearest and bicubic PR is failing in the build, it is not executing the test cases.